### DISCUSSION TITLE
[API Proposal]: "that" syntax for shorter single-line conditionals.

### DISCUSSION THREAD

**Discussion by Glatrix** [dotnet/csharplang#6716] [Language Ideas] [answered]
### Background and motivation

**Background and Motivation** I stumbled across this Idea while trying to cleanup my code. I noticed single-line conditional
statements could be repetitive, especially if the compared object(s) are deep in the class hierarchy.

**Purpose** To fix the problem listed above, while also adding another keyword
that can be used for similar syntax improvements. 

**API**
Its more syntactic sugar than actual API
additions.

**Insight**
- I pulled this Idea out of nowhere, and have not put much thought into it. Because of such, I
don't foresee this idea being Implemented at all, if at any capacity near the base idea. However, with
larger single-line conditionals, maybe even containing nested conditionals, this could save space
and offer better readability. Not much research or planning has gone into this Idea, but I figured it was
worth a shot pitching the Idea.

### API Proposal

```
N/A
```

### API Usage

```csharp
   
    public void Main()
    {
        LayerA A = new LayerA();

        //Both Conditionals will have the same outcome.

        //Normal Conditional
        int normalizedValueA = A.B.C.exampleValue > 0 ? 0 : A.B.C.exampleValue;
        Console.WriteLine(normalizedValueA);

        //Shorter Conditional
        int normalizedValueB = A.B.C.exampleValue > 0 ? 0 : that;
        Console.WriteLine(normalizedValueB);
    }

    class LayerA { public LayerB B = new LayerB(); }
    class LayerB { public LayerC C = new LayerC(); }
    class LayerC{ public int exampleValue = 0; }
```

### Alternative Designs

N / A

### Risks

**Caveats / Risks**
- `that` would hypothetically only take the value present in the first half of the conditional.

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by TahirAhmadov** [comment, marked as answer]
I would like to see this done using the syntax I proposed in https://github.com/dotnet/csharplang/discussions/5404:
```cs
int normalizedValueB = (var temp = A.B.C.exampleValue) > 0 ? 0 : temp;
```
However in conditional contexts like what you have, it's already possible like this:
```cs
int normalizedValueB = A.B.C.exampleValue is var temp && temp > 0 ? 0 : temp;
```

**Reply by Glatrix** [reply]
Ahh, My laziness gets me again.
Anyways, I do like the #5404 proposal. Would love to see this as well.

**Comment by theunrepentantgeek** [comment]
What's so sacrosanct about making this a single line expression?

Why not
``` csharp
var that = A.B.C.exampleValue
int normalizedValueA = that > 0 ? 0 : that;
```

Or, more realistically given that normalization would be something you'd want in a number of places

``` csharp
int normalizedValueA = Normalize(A.B.exampleValue)

// elsewhere
public int Normalize(int value) => value > 0 ? 0 : value;
```

This later form applies DRY in a nice way.

**Reply by Joe4evr** [reply]
Also, in the case of numbers, this operation is basically `Math.Min(value, 0)`.
